File: app/javascript/channels/campaign_channel.js
import consumer from "./consumer"

$(document).on('turbolinks:load', function() {
  consumer.subscriptions.create({
      channel: 'CampaignChannel',
      id: $('#campaign').data('id')
    },

    {
      // Called when the subscription is ready for use on the server
      connected() { },

      // Called when the subscription has been terminated by the server
      disconnected() { },

      // Called when there's incoming data on the websocket for this channel
      received(data) {
        if (!data['dice_roll']) { return; }
        const $diceRolls = $('#dice-rolls');

        while ($diceRolls.find('.notification').length >= 5) {
          $diceRolls.find('.notification').last().remove();
        }
        const $diceRollResult = $(data['dice_roll']);
        $diceRollResult.dismissible();
        $diceRolls.prepend($diceRollResult);
        setTimeout(function () {
          $diceRollResult.fadeOut('normal', function () { $(this).remove(); });
        }, 10000)
      }
    }
  );
});

File: app/channels/campaign_channel.rb
class CampaignChannel < ApplicationCable::Channel
  def subscribed
    stream_from "campaign_#{params[:id]}_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def save_and_broadcast_dice_roll(data)
    # TODO: pass the string to DiceRoll model
    DiceRoll.create(query: data['dice_roll_query'])
  end
end
